### ISSUE TITLE
P.createLanguage v2

### ISSUE THREAD

**Issue by wavebeem** [jneen/parsimmon#210]
Perhaps a few modifications to P.createLanguage could make it more useful. Unfortunately, I don't think there's any good way to do this without breaking backwards compatibility.

https://runkit.com/wavebeem/59cdac70837de0001248bd64

**Comment by MaxGabriel** [comment]
Hi, I just ran into the issue described at #243. I had a parser made with correct language, and then at the last minute received a feature request that necessitated passing arguments to my parsers. This required an annoying refactor away from `createLanguage` that cascaded from a lower level parser to higher level ones. So I'm definitely in favor of supporting that.

**Comment by wavebeem** [comment]
Definitely want to revisit this but I'm a little paralyzed about version 2 at the moment

**Comment by MaxGabriel** [comment]
It's ok! I'm just glad to have a parser combinator library in Javascript (or Typescript for me). Great work on this :D

**Comment by wavebeem** [comment]
Funny to mention that because part of the reason it's been stuck is that I
wanna redo it in TS but it's hard without breaking a lot of old things...
Hah

**Comment by wavebeem** [comment]
@MaxGabriel 

BTW if you have thoughts on how Parsimmon could be better for TS, please please post them in this issue: https://github.com/jneen/parsimmon/issues/230

I've been using TS a lot recently but I never used TS while I was writing parsers, so I could use more thoughts in that department.

**Comment by fresheneesz** [comment]
@wavebeem Could you comment on why you closed this?

**Comment by wavebeem** [comment]
@fresheneesz Per https://github.com/jneen/parsimmon/issues/230 I've been working on a Parsimmon successor that may eventually be Parsimmon v2 (https://www.npmjs.com/package/bread-n-butter)

In that library (bnb), I found that I no longer desired `createLanguage`. The main reason I wanted `createLanguage` in the first place was due to `var` hoisting making recursive parsers hard to write. Now that we can use `let`/`const` in JS, I think it's more straightforward to just throw in a `.lazy` call whenever you have an issues with referencing a variable not yet defined.

So in short, I don't see `createLanguage` having a place going forward.

**Comment by fresheneesz** [comment]
I see, thanks! I'll have to take a deeper look at bnb sometime.

**Comment by wavebeem** [comment]
@fresheneesz If you do get a chance to look, I would love feedback as GitHub issues 😄
